### PR TITLE
Add UI mvc

### LINKED ISSUE #18: implement mvc for UI
the UI code is highly high coupled to the main class. move UI properties and methods to a decoupled pattern (mvc).

### PR DESCRIPTION
**Pull request by kmilo9999**
Reorganized code. Now the logic is split in logic and view layers. All the events are send to the logic layer (VRVolumeAPP) which control the flow to the view class (UIView)
The logic layer still has 900 lines of code, but at least the code is more organized. The VR APP interface when down from 1.9k to 500 loc.
All the features were tested successfully.
Closes #18

### CODE REVIEW

File: src/UI/UIView.cpp [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+                {
+                  tfn_widget_multi[m_selectedVolume].setBlendedHistogram(
+                    m_volumes[m_selectedVolume][active_volume]->getTransferfunction(i),
+                    m_volumes[m_selectedVolume][active_volume2]->getTransferfunction(i), alpha, i);
+                }*/
+
+            }
+            else {
+              /*		tfn_widget_multi[m_selectedVolume].setHistogram(m_volumes[m_selectedVolume][0]->getTransferfunction(i), i);*/
+            }
```
Comment (broarr) [inline src/UI/UIView.cpp:142]: Do you need this if statement still?

File: src/UI/UIView.cpp [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+              tfn_widget[m_selectedVolume].setMinMax(m_volumes[m_selectedVolume][active_volume]->getMin() * alpha + m_volumes[m_selectedVolume][active_volume2]->getMin() * (1.0 - alpha),
+                m_volumes[m_selectedVolume][active_volume]->getMax() * alpha + m_volumes[m_selectedVolume][active_volume2]->getMax() * (1.0 - alpha));
+              if (active_volume < m_volumes[m_selectedVolume].size() && active_volume2 < m_volumes[m_selectedVolume].size())
+                tfn_widget[m_selectedVolume].setBlendedHistogram(m_volumes[m_selectedVolume][active_volume]->getTransferfunction(0), m_volumes[m_selectedVolume][active_volume2]->getTransferfunction(0), alpha);*/
+
+          }
+          else if (numVolumes > 0) {
+            /*tfn_widget[m_selectedVolume].setHistogram(m_volumes[m_selectedVolume][0]->getTransferfunction(0));
+            tfn_widget[m_selectedVolume].setMinMax(m_volumes[m_selectedVolume][0]->getMin(), m_volumes[m_selectedVolume][0]->getMax());*/
+          }
```
Comment (broarr) [inline src/UI/UIView.cpp:166]: Likewise here, do you need this?

**Review by broarr — APPROVED** [review]
LGTM! Thanks for sharing!

**Review by mirestrepo — COMMENTED** [review]
I think there is a lot of commented out code laying around. While I know that some you want to keep, at some point you have to start trusting git ;). Also for the commented out code that you think you must keep, then write an explicit comment why you are keeping it. You do that in few places, but there are many others, that it's not clear why you need that code lying around.

The `VRVolumeApp.h` class seems big. I see the loaders as a clear separate abstraction  but maybe if your menus keep growing, it maybe worth thinking of splitting up the controller a bit base on functionality

File: src/UI/UIView.cpp [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+
+#include <glm/gtc/type_ptr.hpp>
+
+UIView::UIView(VRVolumeApp& controllerApp):m_controller_app(controllerApp), m_multiplier( 1.0f ), m_threshold{ 0.0f },
+m_z_scale{ 0.16f }, m_scale{ 1.0f }, m_slices(256), m_dynamic_slices{ false }, m_renderVolume(true), m_selectedTrnFnc(0),
+m_animated(false), m_ui_frame_controller{ 0.0f }, m_menu_handler(nullptr), m_initialized(false), m_use_transferfunction( false ),
+m_clip_max( 1.0 ), m_clip_min( 0.0 ), m_clip_ypr( 0.0 ), m_clip_pos( 0.0 ), m_useCustomClipPlane( false ), m_rendermethod( 1 ), m_renderchannel( 0 )
+{
+
+}
```
[the remark this answers is not in the document]
Comment (mirestrepo) [reply src/UI/UIView.cpp:16]: I agree, it's hard to read, but I think list initialization [is more efficient](https://isocpp.org/wiki/faq/ctors#init-lists). I am also confused by the mix of brackets....

File: shaders/shader.frag [hunk 4]
```diff
@@ -15,9 +15,10 @@ void main() {
     //phong-shading code in case we need to add it back
     //float ambient = 0.1;
     //float diffuse = clamp(dot(norm, normalize(lightPos - pos)), 0.0, 1.0);
+
```
Comment (mcmcgrath13) [inline shaders/shader.frag:18]: can we rely on git to get back to this code if needed?

File: src/vrapp/VRVolumeApp.cpp [hunk 5]
```diff
@@ -0,0 +1,1027 @@
...
+#include "../../external/msvc/dirent.h"
+#else
+#define OS_SLASH_LOCAL "//"
+#include <dirent.h>
+#endif
+
+#include <glm/gtx/euler_angles.hpp>
+
+
+VRVolumeApp::VRVolumeApp():m_mesh_model(nullptr), m_clip_max{ 1.0f }, m_clip_min{ 0.0f }, m_clip_ypr{ 0.0f }, m_clip_pos{ 0.0 }, m_wasd_pressed(0),
```
Comment (mcmcgrath13) [inline src/vrapp/VRVolumeApp.cpp:33]: the formatting on this makes it a bit hard to read - can it be prettified?

**Review by mcmcgrath13 — COMMENTED** [review]
definitely headed in the right direction!  Some of the indenting looks a bit funny to me - I'd throw it all through an auto-formatter.

File: src/UI/UIView.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kmilo9999) [reply src/UI/UIView.cpp]: Let's call that mix "German notation". Fixed

File: src/UI/UIView.cpp (hunk 1 above)
Comment (kmilo9999) [reply src/UI/UIView.cpp]: Let's keep it for now. I'll remove it in the next PR. There are things I want to test without the need for rolling back specific code.

File: src/UI/UIView.cpp (hunk 2 above)
Comment (kmilo9999) [reply src/UI/UIView.cpp]: Let's keep it for now. I'll remove it in the next PR. There are things I want to test without the need for rolling back specific code.

File: include/vrapp/VRVolumeApp.h [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+  void load_volume(std::vector<std::string> vals, std::promise<Volume*>* promise);
+
+  void load_nrrd_file(std::string& filename);
+
+  void update_frame_state();
+
+  void update_3D_ui();
+  void update_2D_ui();
+
+  virtual void clear_data() ;
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply include/vrapp/VRVolumeApp.h]: I thought I need inheritance, but at the end I didnt use it. Removed it

File: include/vrapp/VRVolumeApp.h [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+  void update_fps(float fps);
+  float get_fps();
+
+  void update_dynamic_slices();
+
+  void do_grab(glm::mat4& pose);
+  
+  void intialize_ui();
+
+  void load_mesh_model();
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply include/vrapp/VRVolumeApp.h:82]: I added a loader class  (VRDataLoader.cpp) to parse the description file. The other loaders logic are a bit tied to the vr app loop.

File: include/vrapp/VRVolumeApp.h [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+
+  void update_frame_state();
+
+  void update_3D_ui();
+  void update_2D_ui();
+
+  virtual void clear_data() ;
+  virtual int get_num_volumes() ;
+  virtual bool data_is_multi_channel() ;
+  virtual void get_min_max(const float frame,float & min, float& max) ;
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply include/vrapp/VRVolumeApp.h]: I formatted the Whole Project

File: shaders/shader.frag (hunk 4 above)
Comment (kmilo9999) [reply shaders/shader.frag]: I Removed it.

File: src/vrapp/VRVolumeApp.cpp (hunk 5 above)
Comment (kmilo9999) [reply src/vrapp/VRVolumeApp.cpp]: I formatted the Whole Project. 
pre...tti..fi..ed ? what's that?

File: src/UI/UIView.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mirestrepo) [reply src/UI/UIView.cpp]: Colombo-German notation 🤣

**Review by mirestrepo — APPROVED** [review]

**Review by broarr — APPROVED** [review]

**Review by mcmcgrath13 — APPROVED** [review]
🤩

**Pull request merged**